alerting/src/main/kotlin/org/opensearch/alerting/workflow/CompositeWorkflowRunner.kt

```diff
@@ -0,0 +1,155 @@
+                    IllegalStateException("Delegate monitor not found ${delegate.monitorId} for the workflow $workflow.id")
+                )
+            if (delegate.chainedMonitorFindings != null) {
+                val chainedMonitor = monitorsById[delegate.chainedMonitorFindings!!.monitorId]
+                    ?: throw AlertingException.wrap(
+                        IllegalStateException("Chained finding monitor not found ${delegate.monitorId} for the workflow $workflow.id")
+                    )
+
+                try {
+                    indexToDocIds = monitorCtx.workflowService!!.getFindingDocIdsByExecutionId(chainedMonitor, executionId)
+                } catch (e: Exception) {
+                    logger.error("Failed to execute workflow. Error: ${e.message}")
+                    return workflowResult.copy(error = AlertingException.wrap(e))
+                }
+            }
+
+            val workflowRunContext = WorkflowRunContext(workflow.id, delegate.chainedMonitorFindings?.monitorId, executionId, indexToDocIds)
+
+            var delegateRunResult: MonitorRunResult<*>?
+            try {
+                delegateRunResult = if (delegateMonitor.isBucketLevelMonitor()) {
+                    BucketLevelMonitorRunner.runMonitor(
+                        delegateMonitor,
+                        monitorCtx,
+                        periodStart,
+                        periodEnd,
+                        dryRun,
+                        workflowRunContext
+                    )
+                } else if (delegateMonitor.isDocLevelMonitor()) {
+                    DocumentLevelMonitorRunner.runMonitor(
+                        delegateMonitor,
+                        monitorCtx,
+                        periodStart,
+                        periodEnd,
+                        dryRun,
+                        workflowRunContext
+                    )
+                } else if (delegateMonitor.isQueryLevelMonitor()) {
+                    QueryLevelMonitorRunner.runMonitor(
+                        delegateMonitor,
+                        monitorCtx,
+                        periodStart,
+                        periodEnd,
+                        dryRun,
+                        workflowRunContext
+                    )
+                } else {
+                    throw AlertingException.wrap(
+                        IllegalStateException("Unsupported monitor type")
+                    )
+                }
+            } catch (ex: Exception) {
+                logger.error("Error executing workflow delegate. Error: ${ex.message}")
+                lastErrorDelegateRun = AlertingException.wrap(ex)
+                continue
+            }
+            if (delegateRunResult != null) resultList.add(delegateRunResult)
+        }
+        logger.debug("Workflow ${workflow.id} in $executionId finished")
+        // Update metadata only if the workflow is not temp
+        if (!isTempMonitor) {
```

Comment:
i hope we are not updating monitor metadata if there was an error in the workflow

Reply:
Check the comment above - need to think how and when to cancel workflow metadata update. In any case, the workflow meatadata are not used like for example metadata in doc level monitors (and their lastRunContext)